File: clients/java/src/main/java/dev/happywakey/HappyWakeyClient.java
package dev.happywakey;
import java.net.URI;
import java.net.http.*;
import java.time.Duration;
import java.util.function.Consumer;

public final class HappyWakeyClient {
  private final URI base; private final String token; private final HttpClient http; private final Consumer<String> telemetry;
  public HappyWakeyClient(URI base, String token, Consumer<String> telemetry) { if (!"https".equals(base.getScheme())) throw new IllegalArgumentException("HTTPS required"); this.base=base; this.token=token; this.telemetry=telemetry; this.http=HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).followRedirects(HttpClient.Redirect.NEVER).build(); }
  public String health() throws Exception { return call("health","GET","/healthz",null,false); }
  public String listAlarms() throws Exception { return call("list_alarms","GET","/v1/alarms",null,true); }
  public String createAlarm(String json) throws Exception { return call("create_alarm","POST","/v1/alarms",json,true); }
  public String transitionOccurrence(String id,String json) throws Exception { return call("transition_occurrence","POST","/v1/occurrences/"+id+"/transitions",json,true); }
  public String pullChanges(String cursor,int limit) throws Exception { return call("pull_changes","GET","/v1/sync/pull?cursor="+cursor+"&limit="+limit,null,true); }
  public String pushChanges(String json) throws Exception { return call("push_changes","POST","/v1/sync/push",json,true); }
  private String call(String operation,String method,String path,String body,boolean auth) throws Exception { if(auth && token.isBlank()) throw new IllegalStateException("Shared Auth bearer token required"); var builder=HttpRequest.newBuilder(base.resolve(path)).timeout(Duration.ofSeconds(10)).header("Accept","application/json"); if(auth) builder.header("Authorization","Bearer "+token); builder.method(method,body==null?HttpRequest.BodyPublishers.noBody():HttpRequest.BodyPublishers.ofString(body)); var response=http.send(builder.build(),HttpResponse.BodyHandlers.ofString()); telemetry.accept("{\"schema\":\"next-loggers/v1\",\"operation\":\""+operation+"\",\"status\":"+response.statusCode()+"}"); if(response.statusCode()<200||response.statusCode()>299) throw new IllegalStateException("Happy Wakey request failed ("+response.statusCode()+")"); return response.body(); }
}

File: clients/kotlin/src/main/kotlin/dev/happywakey/HappyWakeyClient.kt
package dev.happywakey
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

class HappyWakeyClient(private val base: URI, private val token: String, private val telemetry: (String, Int) -> Unit) {
  private val http = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER).build()
  init { require(base.scheme == "https") { "HTTPS required" } }
  fun health() = call("health", "GET", "/healthz", null, false)
  fun listAlarms() = call("list_alarms", "GET", "/v1/alarms")
  fun createAlarm(json: String) = call("create_alarm", "POST", "/v1/alarms", json)
  fun transitionOccurrence(id: String, json: String) = call("transition_occurrence", "POST", "/v1/occurrences/$id/transitions", json)
  fun pullChanges(cursor: String = "0", limit: Int = 100) = call("pull_changes", "GET", "/v1/sync/pull?cursor=$cursor&limit=$limit")
  fun pushChanges(json: String) = call("push_changes", "POST", "/v1/sync/push", json)
  private fun call(operation:String,method:String,path:String,body:String?=null,auth:Boolean=true):String { if(auth) require(token.isNotBlank()){"Shared Auth bearer token required"}; val builder=HttpRequest.newBuilder(base.resolve(path)).header("Accept","application/json"); if(auth) builder.header("Authorization","Bearer $token"); builder.method(method,body?.let{HttpRequest.BodyPublishers.ofString(it)}?:HttpRequest.BodyPublishers.noBody()); val response=http.send(builder.build(),HttpResponse.BodyHandlers.ofString()); telemetry("next-loggers/v1:happy_wakey.client.request:$operation",response.statusCode()); check(response.statusCode() in 200..299){"Happy Wakey request failed (${response.statusCode()})"}; return response.body() }
}

